<a href="https://colab.research.google.com/github/Putri-andini-rpe/Pemrograman-Dasar/blob/main/Pertemuan_10_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Tulis kode program diatas untuk melihat dan menghubungkan dengan google drive "Pertemuan 10"yang akan ditautkan. Jika sudah already, artinya drive sudah terhubung.


In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


In [ ]:
customers_df =pd.read_csv("/content/drive/MyDrive/Pemrograman Dasar/Pertemuan 10/customers.csv")
customers_df.head()

customer_id customer_name             gender  age  \
0            1       fulan 1             Female   30   
1            2       fulan 2  Prefer not to say   69   
2            3       fulan 3  Prefer not to say   59   
3            4       fulan 4  Prefer not to say   67   
4            5       fulan 5  Prefer not to say   30   

                     home_address  zip_code               city  \
0  8606 Victoria TerraceSuite 560      5464      Johnstonhaven   
1      8327 Kirlin SummitApt. 461      8223    New Zacharyfort   
2       269 Gemma SummitSuite 109      5661           Aliburgh   
3       743 Bailey GroveSuite 141      1729  South Justinhaven   
4         48 Hyatt ManorSuite 375      4032     Griffithsshire   

                          state    country  
0            Northern Territory  Australia  
1               South Australia  Australia  
2  Australian Capital Territory  Australia  
3                    Queensland  Australia  
4                    Queensland  Australia

Memuat tabel customers

In [ ]:
orders_df =pd.read_csv("/content/drive/MyDrive/Pemrograman Dasar/Pertemuan 10/orders.csv")
orders_df.head()

order_id  customer_id  payment order_date delivery_date
0         1           64    30811  2021-8-30    2021-09-24
1         2          473    50490   2021-2-3    2021-02-13
2         3          774    46763  2021-10-8    2021-11-03
3         4          433    39782   2021-5-6    2021-05-19
4         5          441    14719  2021-3-23    2021-03-24

Memuat tabel orders

In [ ]:
products_df =pd.read_csv("/content/drive/MyDrive/Pemrograman Dasar/Pertemuan 10/products.csv")
products_df.head()

product_id product_type  product_name size colour  price  quantity  \
0           0        Shirt  Oxford Cloth   XS    red    114        66   
1           1        Shirt  Oxford Cloth    S    red    114        53   
2           2        Shirt  Oxford Cloth    M    red    114        54   
3           3        Shirt  Oxford Cloth    L    red    114        69   
4           4        Shirt  Oxford Cloth   XL    red    114        47   

                                    description  
0  A red coloured, XS sized, Oxford Cloth Shirt  
1   A red coloured, S sized, Oxford Cloth Shirt  
2   A red coloured, M sized, Oxford Cloth Shirt  
3   A red coloured, L sized, Oxford Cloth Shirt  
4  A red coloured, XL sized, Oxford Cloth Shirt

Memuat tabel product

In [ ]:
sales_df =pd.read_csv("/content/drive/MyDrive/Pemrograman Dasar/Pertemuan 10/sales.csv")
sales_df.head()


sales_id  order_id  product_id  price_per_unit  quantity  total_price
0         0         1         218             106         2        212.0
1         1         1         481             118         1        118.0
2         2         1           2              96         3        288.0
3         3         1        1002             106         2        212.0
4         4         1         691             113         3        339.0

Memuat tabel sales

In [ ]:
customers_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1007 entries, 0 to 1006
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customer_id    1007 non-null   int64 
 1   customer_name  1007 non-null   object
 2   gender         989 non-null    object
 3   age            1007 non-null   int64 
 4   home_address   1007 non-null   object
 5   zip_code       1007 non-null   int64 
 6   city           1007 non-null   object
 7   state          1007 non-null   object
 8   country        1007 non-null   object
dtypes: int64(3), object(6)
memory usage: 70.9+ KB


Assessing Data sekarang masuk ke tahap kedua dalam proses data wrangling yaitu assessing data. Menilai data customers_df yaitu data pertama yang akan kita nilai adalah customers_df. Proses ini dapat dilakukan dengan menggunakan method info(). Jika di perhatikan, tidak ada masalah dengan tipe data dari seluruh kolom tersebut. Akan tetapi, terdapat sedikit perbedaan pada jumlah data pada kolom gender. Hal ini menunjukkan adanya missing values pada kolom gender. Lalu jalankan kode berikutnya.

In [ ]:
customers_df.isna().sum()

customer_id       0
customer_name     0
gender           18
age               0
home_address      0
zip_code          0
city              0
state             0
country           0
dtype: int64

Terdapat 18 data yang hilang atau missing value pada kolom gender. Hal ini yang akan di tangani pada tahap data cleaning. Masalah selanjutnya yang harus diperiksa adalah duplikasi data. Ini menggunakan method df.duplicated(). Untuk melihat output ini, kita bisa menjalankan kode program berikut.

In [ ]:
print("Jumlah duplikasi: ", customers_df.duplicated().sum())

Jumlah duplikasi:  6


Ketika menjalankan kode di atas, akan menemukan 6 data yang terduplikat. Pada tahap data cleaning, nanti akan menghilangkan semua duplikasi itu. Selanjutnya kita memeriksa parameter statistik dari kolom numerik yang terdapat dalam customers_df. Untuk memudahkan pemeriksaan, kita menggunakan method describe(). Method akan menampilkan ringkasan parameter statistik (mean, median, dll) dari kolom numerik pada sebuah DataFrame. Berikut contoh penggunaannya.

In [ ]:
customers_df.describe()

customer_id          age     zip_code
count  1007.000000  1007.000000  1007.000000
mean    501.726912    50.929494  5012.538232
std     288.673238    30.516299  2885.836112
min       1.000000    20.000000     2.000000
25%     252.500000    34.000000  2403.500000
50%     502.000000    50.000000  5087.000000
75%     751.500000    65.000000  7493.500000
max    1000.000000   700.000000  9998.000000

Jika diperhatikan, terdapat keanehan pada nilai maksimum yang ada pada kolom age. Kemungkinan besar ini terjadi karena adanya inaccurate value pada kolom tersebut. Masalah ini juga akan dibersihkan ke dalam tahap cleaning data.

In [ ]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   order_id       1000 non-null   int64 
 1   customer_id    1000 non-null   int64 
 2   payment        1000 non-null   int64 
 3   order_date     1000 non-null   object
 4   delivery_date  1000 non-null   object
dtypes: int64(3), object(2)
memory usage: 39.2+ KB


Digunakan untuk menilai data dan memeriksa tipe data dari tiap kolom menggunakan method. Jika diperhatikan, jumlah data pada orders_df telah lengkap atau dengan kata lain tidak ada missing value didalamnya. Akan tetapi, terdapat kesalahan tipe data untuk kolom order_date dan delivery_date. Kedua kolom tersebut harusnya direpresentasikan sebagai tipe data datetime. Berikutnya akan memeriksa duplikasi dan ringkasan parameter statistik dari kolom numerik pada orders_df

In [ ]:
print("Jumlah duplikasi: ",orders_df.duplicated().sum())

Jumlah duplikasi:  0


In [ ]:
orders_df.describe()

order_id  customer_id       payment
count  1000.000000  1000.000000   1000.000000
mean    500.500000   506.640000  33972.936000
std     288.819436   277.115502  14451.609047
min       1.000000     1.000000  10043.000000
25%     250.750000   275.250000  21329.250000
50%     500.500000   515.000000  33697.500000
75%     750.250000   737.250000  46249.000000
max    1000.000000  1000.000000  59910.000000

Jika diperhatikan, tidak ada keanehan data pada hasil tersebut. Ini menunjukkan tidak adanya duplikasi dan keanehan nilai pada orders_df

In [ ]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1266 entries, 0 to 1265
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   product_id    1266 non-null   int64 
 1   product_type  1266 non-null   object
 2   product_name  1266 non-null   object
 3   size          1266 non-null   object
 4   colour        1266 non-null   object
 5   price         1266 non-null   int64 
 6   quantity      1266 non-null   int64 
 7   description   1266 non-null   object
dtypes: int64(3), object(5)
memory usage: 79.3+ KB


Berdasarkan hasil data tersebut,dapat disimpulkan bahwa tidak terdapat masalah pada tipe data tiap kolom data product_df. Selain itu, jumlah data juga lengkap sehingga tidak ada missing value didalamnya.

In [ ]:
print("Jumlah duplikasi: ", products_df.duplicated().sum())

Jumlah duplikasi:  6


In [ ]:
products_df.describe()

product_id        price     quantity
count  1266.000000  1266.000000  1266.000000
mean    627.926540   105.812006    60.138231
std     363.971586     9.715611    11.682791
min       0.000000    90.000000    40.000000
25%     313.250000    95.250000    50.000000
50%     626.500000   109.000000    60.000000
75%     942.750000   114.000000    70.000000
max    1259.000000   119.000000    80.000000

Berdasarkan data diatas, dapat dillihat bahwa terdapat 6 data yang terduplikasi pada product_df. Pada tahap cleaning data, kita akan menghapus duplikasi tersebut. Menilai data sales_df, Data terakhir yang harus kita nilai ialah sales_df sebagai permulaan, gunakan method info() dan untuk memeriksa tipe data dan jumlah data yang ada didalam sales_df

In [ ]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   sales_id        5000 non-null   int64  
 1   order_id        5000 non-null   int64  
 2   product_id      5000 non-null   int64  
 3   price_per_unit  5000 non-null   int64  
 4   quantity        5000 non-null   int64  
 5   total_price     4981 non-null   float64
dtypes: float64(1), int64(5)
memory usage: 234.5 KB


Berdasarkan hasil di atas, tidak ada masalah pada tipe data tiap kolom dalam sales_df. Namun terdapat keanehan pada jumlah data pada kolom total_price. Ini menunjukkab adanya missing value pada kolom tersebut.

In [ ]:
sales_df.isna().sum()

sales_id           0
order_id           0
product_id         0
price_per_unit     0
quantity           0
total_price       19
dtype: int64

Kode di atas akan menunjukkan jumlah missing value dari setiap kolom dalam sales_df. Berdasarkan hasil di atas, dapat disimpulkan bahwa terdapat 19 missing value pada kolom total_price. Hal ini akan kita bersihkan pada tahap cleaning data.

selanjutnya ialah memeriksa duplikasi dan ringkasan parameter statistik dari sales_df

In [ ]:
print("Jumlah duplikasi: ", sales_df.duplicated().sum())

Jumlah duplikasi:  0


In [ ]:
sales_df.describe()

sales_id     order_id   product_id  price_per_unit    quantity  \
count  5000.000000  5000.000000  5000.000000     5000.000000  5000.00000   
mean   2499.500000   503.038200   634.053200      103.501600     1.99240   
std    1443.520003   285.964418   363.255794        9.195004     0.80751   
min       0.000000     1.000000     1.000000       90.000000     1.00000   
25%    1249.750000   258.000000   323.000000       95.000000     1.00000   
50%    2499.500000   504.500000   635.000000      102.000000     2.00000   
75%    3749.250000   749.000000   951.000000      112.000000     3.00000   
max    4999.000000   999.000000  1259.000000      119.000000     3.00000   

       total_price  
count  4981.000000  
mean    206.307368  
std      86.352449  
min      90.000000  
25%     112.000000  
50%     204.000000  
75%     285.000000  
max     357.000000

Hasil di atas menunjukkan bahwa tidak ada duplikasi pada sales_df. Selain itu, ia juga menunjukkan tidak terdapat keanehan dalam ringkasan parameter statistik dari sales_df. Sampai tahap ini, kita telah berhasil mengidentifikasi berbagai masalah pada data yang telah kita kumpulkan. Berikut merupakan rangkumannya ▶ Terdapat 18 missing values pada kolom gender, terdapat 6 data yang duplikat, terdapat inaccurate value pada kolom age, terdapat kesalahan tipe data untuk kolom order_date & delivery_date, terdapat 6 data yang duplikat, dan terdapat 19 missing value pada kolom total_price.
Membersihkan Data customers_df Berdasarkan hasil proses assessing data, diketahui bahwa terdapat tiga masalah yang dijumpai dalam customer_df, yaitu duplicate data, missing value, dan inaccurate value. Menghilangkan duplicate data

untuk melakukan hal ini, kita dapat memanfaatkan method drop_duplicates(). Berikut merupakan kode untuk menghapus duplikasi pada customer_df.

In [ ]:
customers_df.drop_duplicates(inplace=True)

Setelah menjalankan kode di atas, periksa kembali apakah masih terdapat duplikasi pada data tersebut dengan menjalankan kode berikut.

In [ ]:
print("Jumlah duplikasi: ", customers_df.duplicated().sum())

Jumlah duplikasi:  0


Jika proses penghapusan duplikasi berjalan lancar, kode di atas akan menghasilkan keluaran yang
menunjukkan tidak adanya duplikasi pada customers_df.
Masalah berikutnya yang harus kita tangani ialah missing value pada kolom gender. Nah, secara umum terdapat tiga metode untuk mengatasi missing value yaitu dropping, imputation, dan interpolation. Untuk menentukan metode mana yang akan digunakan, kita perlu melihat data yang mengandung missing value tersebut menggunakan teknik filtering seperti berikut.

In [ ]:
customers_df[customers_df.gender.isna()]

customer_id customer_name gender  age                     home_address  \
38            39      fulan 39    NaN   80  7440 Cameron Estate DrSuite 628   
167          168     fulan 168    NaN   27         2781 Berge MallSuite 452   
322          322     fulan 322    NaN   30        593 Becker CircleApt. 333   
393          393     fulan 393    NaN   34          5158 Levi HillSuite 531   
442          442     fulan 442    NaN   26           5157 Feil RoadApt. 633   
722          720     fulan 720    NaN   40         31 Jordan ParadeApt. 400   
745          743     fulan 743    NaN   57   09 Christopher StreetSuite 967   
773          771     fulan 771    NaN   74     7367 Wright JunctionApt. 773   
798          795     fulan 795    NaN   49          487 Summer MewsApt. 874   
801          798     fulan 798    NaN   56           27 Aiden KnollApt. 875   
825          822     fulan 822    NaN   59        41 Jenkins KnollSuite 438   
859          855     fulan 855    NaN   55       603 O'keefe KnollSuite 782   
863          859     fulan 859    NaN   38            32 Isla GroveApt. 078   
914          909     fulan 909    NaN   62    976 Murray Station StApt. 036   
934          929     fulan 929    NaN   68           394 Lily HillSuite 153   
948          943     fulan 943    NaN   64       3117 Heller PlaceSuite 149   
952          946     fulan 946    NaN   24      8227 Nicholas HillSuite 150   
994          988     fulan 988    NaN   35   1130 Turner Estate DrSuite 925   

     zip_code                   city                         state    country  
38       4622  North Victoriachester            Northern Territory  Australia  
167      1975         North Leoburgh             Western Australia  Australia  
322      1640             Jacobiview             Western Australia  Australia  
393      1474             Johnsburgh                    Queensland  Australia  
442      7249             Port Chloe               New South Wales  Australia  
722      1380             West Henry               South Australia  Australia  
745      6226         Lake Lukemouth             Western Australia  Australia  
773      8882               Kuhntown                      Victoria  Australia  
798      1712            East Hayden  Australian Capital Territory  Australia  
801      6531               Port Sam  Australian Capital Territory  Australia  
825      2588        Lake Andrewport               South Australia  Australia  
859      8822        Port Dylanmouth                      Tasmania  Australia  
863      7711            Rosechester               New South Wales  Australia  
914      3227               Langfort                      Tasmania  Australia  
934      2353             Beahanfurt            Northern Territory  Australia  
948       822           North Elijah               South Australia  Australia  
952       115           South Jasper                    Queensland  Australia  
994      9386              New Harry             Western Australia  Australia

Kode di atas hanya akan menampilkan baris data yang memenuhi kondisi customers_df.gender.isna() atau dengan kata lain ia akan menampilkan baris data yang mengandung missing value pada kolom gender

Berdasarkan gambar di atas, dapat dilihat bahwa baris data tersebut masih mengandung banyak informasi penting sehingga sayang jika langsung dibuang. Oleh karena itu, pada kasus ini, kita akan menggunakan metode imputation untuk menangani missing value.

Gunakanlah method value_counts()untuk mengidentifikasi nilai yang dominan.

In [ ]:
customers_df.gender.value_counts()

gender
Prefer not to say    725
Male                 143
Female               115
Name: count, dtype: int64

Berdasarkan hasil di atas, dapat diketahui bahwa nilai yang paling dominan dalam kolom gender ialah “Prefer not to say”. Nilai inilah yang selanjutnya akan kita gunakan sebagai pengganti missing value. Proses penggantian ini dapat dilakukan menggunakan method fillna()seperti contoh berikut.

Untuk memastikan proses di atas berjalan dengan semestinya, kita bisa menjalankan kembali kode untuk mengidentifikasi missing value seperti berikut.

In [ ]:
customers_df.isna().sum()

customer_id       0
customer_name     0
gender           18
age               0
home_address      0
zip_code          0
city              0
state             0
country           0
dtype: int64

sekarang kita akan mengatasi masalah inaccurate value pada kolom age. Sebagai awal, kita perlu melihat data baris data yang mengandung inaccurate value tersebut (baris dengan nilai age maksimum). Hal ini dilakukan menggunakan teknik filter seperti contoh kode berikut.

In [ ]:
customers_df[customers_df.age == customers_df.age.max()]

customer_id customer_name             gender  age  \
967          961     fulan 961  Prefer not to say  700   

                   home_address  zip_code        city            state  \
967  29 Farrell ParadeSuite 818      6528  New Joseph  South Australia   

       country  
967  Australia

Kode di atas akan menampilkan baris data yang memiliki nilai age maksimum.

Berdasarkan data tersebut, kita bisa berasumsi bahwa inaccurate value tersebut terjadi karena human error sehingga kelebihan memasukkan nilai nol. Oleh karena itu, gantilah dengan nilai 70 . Proses ini dilakukan dengan memanfaatkan method replace()seperti contoh berikut.

In [ ]:
customers_df.age.replace(customers_df.age.max(), 70, inplace=True)

/tmp/ipython-input-4147959128.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  customers_df.age.replace(customers_df.age.max(), 70, inplace=True)


In [ ]:
customers_df[customers_df.age == customers_df.age.max()]

customer_id customer_name             gender  age  \
215          216     fulan 216  Prefer not to say  500   

               home_address  zip_code       city               state  \
215  038 Haley MewsApt. 810      3991  Bayertown  Northern Territory   

       country  
215  Australia

ternyata masih ada invalid value lain yang terdapat dalam kolom age.

Penyebab kesalahan ini kemungkinan sama dengan sebelumnya, yaitu human error yang kelebihan memasukkan nilai nol. Untuk menangani hal ini, kita akan menggantinya dengan nilai 50.

In [ ]:
customers_df.age.replace(customers_df.age.max(), 50, inplace=True)

/tmp/ipython-input-606824494.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  customers_df.age.replace(customers_df.age.max(), 50, inplace=True)


In [ ]:
customers_df[customers_df.age == customers_df.age.max()]

customer_id customer_name             gender  age  \
34            35      fulan 35               Male   80   
38            39      fulan 39                NaN   80   
142          143     fulan 143  Prefer not to say   80   
154          155     fulan 155               Male   80   
170          171     fulan 171  Prefer not to say   80   
174          175     fulan 175  Prefer not to say   80   
181          182     fulan 182               Male   80   
235          236     fulan 236  Prefer not to say   80   
424          424     fulan 424  Prefer not to say   80   
438          438     fulan 438  Prefer not to say   80   
495          495     fulan 495               Male   80   
524          524     fulan 524  Prefer not to say   80   
598          597     fulan 597  Prefer not to say   80   
703          702     fulan 702  Prefer not to say   80   
734          732     fulan 732  Prefer not to say   80   

                          home_address  zip_code                   city  \
34   3168 Bartoletti CrescentSuite 878       593             Port Lucas   
38     7440 Cameron Estate DrSuite 628      4622  North Victoriachester   
142        87 Hahn Station StSuite 943      8254               New Ella   
154           85 Charles MallSuite 424      7841            Port Sophia   
170              16 Boyer WaySuite 018      6226          South Gabriel   
174             57 David MallSuite 596      3129                Haneton   
181           32 Thomas CrestSuite 753      2147             Archiefurt   
235            64 Phillips RunApt. 722      2752            East Sienna   
424            03 Grant CrestSuite 228      5743      Lake Charliemouth   
438          569 Alyssa IslandApt. 766      6893             Hunterland   
495          1798 Schultz MewsApt. 704      9409           Lake Caitlin   
524            0358 Ashley WayApt. 506      3345     Port Sebastianland   
598          78 O'hara CircuitApt. 096      1966       Lake Michaelbury   
703              61 Bosco WaySuite 775       685        Thompsonborough   
734    905 Stephanie BoulevardApt. 217      8114          Port Aaronton   

                  state    country  
34           Queensland  Australia  
38   Northern Territory  Australia  
142            Victoria  Australia  
154     New South Wales  Australia  
170   Western Australia  Australia  
174     South Australia  Australia  
181          Queensland  Australia  
235     New South Wales  Australia  
424   Western Australia  Australia  
438   Western Australia  Australia  
495     New South Wales  Australia  
524          Queensland  Australia  
598     South Australia  Australia  
703   Western Australia  Australia  
734          Queensland  Australia

Pastikan lagi bahwa tidak ada data yang inaccurate pada kolom age

In [ ]:
customers_df.describe()

customer_id          age     zip_code
count  1001.000000  1001.000000  1001.000000
mean    500.942058    49.874126  5000.693307
std     289.013599    17.644663  2886.084454
min       1.000000    20.000000     2.000000
25%     251.000000    34.000000  2398.000000
50%     501.000000    50.000000  5079.000000
75%     751.000000    65.000000  7454.000000
max    1000.000000    80.000000  9998.000000

Berdasarkan hasil tersebut dapat dilihat bahwa kolom age memiliki nilai maksimum yang cukup. Selain itu, jika Ada perhatikan, nilai mean dan standard deviation ikut berubah setelah kita menangani inaccurate value tersebut.

Membersihkan Data orders_df Sekarang kita telah menyelesaikan semua permasalahan yang ada pada customers_df. Selanjutnya, kita akan mengatasi permasalahan pada orders_df. Berdasarkan proses assessing data sebelumnya, diketahui bahwa terdapat kesalahan tipe data untuk kolom order_date & delivery_date.

Untuk mengatasi masalah ini, kita akan mengganti tipe data pada kolom order_date & delivery_date menjadi datetime. Proses ini dapat dilakukan menggunakan function to_datetime() yang disediakan oleh library pandas

In [ ]:
datetime_columns = ["order_date", "delivery_date"]

for column in datetime_columns:
  orders_df[column] = pd.to_datetime(orders_df[column])

Kode di atas akan mengubah tipe data pada kolom order_date dan delivery_date menjadi datetime. Untuk memastikan hal ini berjalan sesuai harapan, periksa kembali tipe data tersebut menggunakan method info()

In [ ]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   order_id       1000 non-null   int64 
 1   customer_id    1000 non-null   int64 
 2   payment        1000 non-null   int64 
 3   order_date     1000 non-null   object
 4   delivery_date  1000 non-null   object
dtypes: int64(3), object(2)
memory usage: 39.2+ KB


Jika semua tahapan berjalan sesuai harapanm, Maka tipe data kedua kolom sudah berganti menjadi datetime. Data berikutnya yang akan bersihkan ialah product_df. Sesuai hasil assessing data sebelumnya, kita mengetahui bahwa terdapat 6 data yang terduplikasi pada product_df. Untuk mengatasi hal ini, kita perlu membuang data yang sama tersebut menggunakan method drop_duplicates() seperti contoh berikut.

In [ ]:
products_df.drop_duplicates(inplace=True)

Kode di atas akan menghapus semua data yang duplikat. Untuk memastikan kode tersebut berjalan sesuai harapan, jalankanlah kode berikut.

In [ ]:
print("Jumlah duplikasi: ", products_df.duplicated().sum())

Jumlah duplikasi:  0


Jika proses penghapusan data yang duplikat berjalan dengan lancar, kode di atas akan menghasilkan keluaran seperti berikut “Jumlah duplikasi: 0”. Data selanjutnya yang perlu bersihkan ialah sales_df. Berdasarkan hasil penilain data sebelumnya, diketahui bahwa terdapat 19 missing value pada kolom total_price. Untuk mengetahui proses penanganan missing value yang paling sesuai, kita perlu melihat terlebih dahulu baris data yang mengandung missing value tersebut.

In [ ]:
sales_df[sales_df.total_price.isna()]

sales_id  order_id  product_id  price_per_unit  quantity  total_price
9            9         2        1196             105         1          NaN
121        121        27        1027              90         3          NaN
278        278        63         360              94         2          NaN
421        421        95        1091             115         1          NaN
489        489       108        1193             105         3          NaN
539        539       117         405             119         2          NaN
636        636       134         653              93         3          NaN
687        687       145        1138             102         1          NaN
854        854       177          64             104         1          NaN
1079      1079       222         908              94         3          NaN
1193      1193       248        1121             102         2          NaN
1313      1313       272         826             117         1          NaN
1548      1548       316         103             118         3          NaN
1688      1688       345         428             107         1          NaN
1775      1775       359         694             113         2          NaN
1902      1902       381        1218             105         3          NaN
2025      2025       408         611             112         3          NaN
2164      2164       436         583             100         3          NaN
2347      2347       476         696             113         2          NaN

Kode di atas akan menampilkan semua baris data yang memiliki missing value pada kolom total_price seperti pada gambar di atas. Berdasarkan tampilan data tersebut, kita menemukan bahwa nilai total_price merupakan hasil perkalian antara price_per_unit dan quantity.

Kita dapat menggunakan pola ini untuk menangani missing value pada kolom total_price. Berikut merupakan contoh penerapan kode untuk melakukannya.

In [ ]:
sales_df["total_price"] = sales_df["price_per_unit"] * sales_df["quantity"]

Kode di atas akan mengatasi seluruh missing value serta memastikan nilai pada kolom total_price telah sesuai. Untuk memastikan hal ini, Anda dapat memeriksa kembali jumlah missing value pada sales_df menggunakan kode berikut.

In [ ]:
sales_df.isna().sum()

sales_id          0
order_id          0
product_id        0
price_per_unit    0
quantity          0
total_price       0
dtype: int64

Jika proses sebelumnya berjalan lancar, Anda akan menemukan hasil seperti data di atas. Tahapan di atas adalah Proses data wrangling untuk menyiapkan data sebelum dianalisis. Setelah memahami semua teori tentang proses EDA, kini saatnya kita menerapkannya dalam sebuah proyek analisis data sederhana.

In [ ]:
customers_df.describe(include="all")

customer_id customer_name             gender          age  \
count   1001.000000          1001                983  1001.000000   
unique          NaN          1000                  3          NaN   
top             NaN     fulan 943  Prefer not to say          NaN   
freq            NaN             2                725          NaN   
mean     500.942058           NaN                NaN    49.874126   
std      289.013599           NaN                NaN    17.644663   
min        1.000000           NaN                NaN    20.000000   
25%      251.000000           NaN                NaN    34.000000   
50%      501.000000           NaN                NaN    50.000000   
75%      751.000000           NaN                NaN    65.000000   
max     1000.000000           NaN                NaN    80.000000   

                      home_address     zip_code        city            state  \
count                         1001  1001.000000        1001             1001   
unique                        1000          NaN         961                8   
top     3117 Heller PlaceSuite 149          NaN  East Aidan  South Australia   
freq                             2          NaN           3              140   
mean                           NaN  5000.693307         NaN              NaN   
std                            NaN  2886.084454         NaN              NaN   
min                            NaN     2.000000         NaN              NaN   
25%                            NaN  2398.000000         NaN              NaN   
50%                            NaN  5079.000000         NaN              NaN   
75%                            NaN  7454.000000         NaN              NaN   
max                            NaN  9998.000000         NaN              NaN   

          country  
count        1001  
unique          1  
top     Australia  
freq         1001  
mean          NaN  
std           NaN  
min           NaN  
25%           NaN  
50%           NaN  
75%           NaN  
max           NaN

Berdasarkan rangkuman parameter statistik di atas, kita akan memperoleh informasi jumlah pelanggan sebanyak 1001 orang yang berumur di kisaran antara 20 hingga 80 tahun dengan rata-rata umur sebesar 49.87 tahun dengan standar deviasinya sebesar 17.64 tahun. Informasi ini tentunya dapat memberikan gambaran bahwa segmentasi pelanggan yang kita miliki cukup luas mulai dari remaja hingga lanjut usia.

Sekarang coba kita lihat demografi pelanggan berdasarkan jenis kelamin (gender). Untuk melakukan ini, kita akan menggunakan method groupby() yang diikuti dengan method agg(). Berikut kode untuk melakukannya.

In [ ]:
customers_df.groupby(by="gender").agg({
    "customer_id": "nunique",
    "age": ["max", "min", "mean", "std"]
})

customer_id age                          
                      nunique max min       mean        std
gender                                                     
Female                    115  79  20  49.147826  16.646607
Male                      143  80  20  51.230769  18.462635
Prefer not to say         725  80  20  49.748966  17.658998

Kode di atas akan menghasilkan sebuah pivot table. Ia memuat informasi jumlah pelanggan (nunique dari customer_id) serta parameter umur yang dikelompokkan berdasarkan jenis kelamin.

Berdasarkan pivot table di atas, dapat diketahui bahwa pelanggan yang kita miliki didominasi oleh jenis kelamin prefer not to say. Di lain sisi, distribusi umur mereka ternyata cukup mirip yaitu berkisar antara 20 hingga 80 tahun.

Selanjutnya, kita coba melihat persebaran jumlah pelanggan berdasarkan kota (city) dan negara bagian (state). Untuk melakukannya kita juga akan menggunakan method groupby(). Selain itu, agar hasilnya lebih mudah untuk dilihat, kita akan mengurutkan nilainya menggunakan method sort_values() secara descending.

In [ ]:
customers_df.groupby(by="city").customer_id.nunique().sort_values(ascending=False)

city
East Sophia     3
New Ava         3
East Aidan      3
Lake Sienna     2
Jademouth       2
               ..
Justinport      1
Kaiberg         1
Kaifort         1
Kaihaven        1
Johnsonhaven    1
Name: customer_id, Length: 961, dtype: int64

In [ ]:
customers_df.groupby(by="state").customer_id.nunique().sort_values(ascending=False)

state
South Australia                 139
Queensland                      134
New South Wales                 132
Northern Territory              125
Western Australia               124
Australian Capital Territory    121
Victoria                        121
Tasmania                        104
Name: customer_id, dtype: int64

Berdasarkan hasil di atas, dapat dilihat bahwa persebaran pelanggan kita cukup merata pada setiap kota dan negara bagian. Pelanggan kita paling banyak berada di kota East Aidan, East Sophia, dan New Ava dengan jumlah masing-masing tiga pelanggan. Selain itu, pelanggan kita paling banyak berasal dari negara bagian South Australia. Sekarang kita telah memiliki cukup banyak informasi terkait data customers_df. Selanjutnya kita akan melakukan eksplorasi terhadap data orders_df

Data kedua yang akan kita eksplor adalah orders_df. Ia mengandung berbagai informasi terkait sebuah order yang terdiri dari order_id, customer_id, order_date, dan delivery_date

Berdasarkan informasi tersebut, kita bisa membuat sebuah kolom baru untuk memuat waktu pengiriman tiap order. Untuk melakukannya, kita perlu menghitung selisih antar delivery_date dan order_date serta menyimpannya sebagai delivery_time. Selanjutnya, kita akan menggunakan method apply() untuk melakukan sebuah operasi terhadap setiap elemen dalam sebuah kolom DataFrame atau Series (bentuk satu dimensi dari DataFrame).

Operasi yang akan kita lakukan ialah menghitung jumlah detik dari delivery_time menggunakan method total_seconds(). selanjutnya diubah ke dalam satuan hari (dibagi 86400) dan diambil bilangan bulatnya

In [ ]:
orders_df['order_date'] = pd.to_datetime(orders_df['order_date'])
orders_df['delivery_date'] = pd.to_datetime(orders_df['delivery_date'])
delivery_time = orders_df["delivery_date"] - orders_df["order_date"]
delivery_time = delivery_time.apply(lambda x: x.total_seconds())
orders_df["delivery_time"] = round(delivery_time/86400)

kode di atas, akan menemukan kolom baru pada DataFrame orders_df

In [ ]:
orders_df.describe(include="all")

order_id  customer_id       payment order_date delivery_date
count   1000.000000  1000.000000   1000.000000       1000          1000
unique          NaN          NaN           NaN        291           305
top             NaN          NaN           NaN  2021-6-20    2021-08-09
freq            NaN          NaN           NaN         10             9
mean     500.500000   506.640000  33972.936000        NaN           NaN
std      288.819436   277.115502  14451.609047        NaN           NaN
min        1.000000     1.000000  10043.000000        NaN           NaN
25%      250.750000   275.250000  21329.250000        NaN           NaN
50%      500.500000   515.000000  33697.500000        NaN           NaN
75%      750.250000   737.250000  46249.000000        NaN           NaN
max     1000.000000  1000.000000  59910.000000        NaN           NaN

Berdasarkan hasil tersebut, dapat diketahui bahwa rata-rata waktu pengiriman sebesar 14 hari dengan nilai maksimum sebesar 27 hari dan nilai minimum sebesar 1 hari.

Jika diperhatikan, pada data orders_df terdapat kolom yang berisi informasi terkait customer id pelanggan yang pernah melakukan order. Informasi ini bisa kita gunakan untuk mengidentifikasi pelanggan yang belum pernah melakukan order.

Untuk melakukan hal ini, kita bisa membuat sebuah kolom baru bernama “status” pada data customers_df. Kolom tersebut memiliki nilai “Active” untuk pelanggan yang pernah melakukan order setidaknya sekali dan sebaliknya bernilai “Non Active” untuk pelanggan yang belum pernah melakukan order sama sekali

In [ ]:
customer_id_in_orders_df =  orders_df.customer_id.tolist()
customers_df["status"] = customers_df["customer_id"].apply(lambda x: "Active" if x in customer_id_in_orders_df else "Non Active")
customers_df.sample(5)

customer_id customer_name             gender  age  \
602          601     fulan 601               Male   55   
497          497     fulan 497  Prefer not to say   31   
341          341     fulan 341             Female   34   
84            85      fulan 85  Prefer not to say   27   
930          925     fulan 925  Prefer not to say   70   

                         home_address  zip_code             city  \
602      010 Cartwright TrackApt. 920      1625    Sebastianbury   
497      0487 Jordan TerraceSuite 509      9290  North Aidentown   
341           003 Bruen DriveApt. 633      6522     Jasmineville   
84         41 Feeney ParkwaySuite 074      4401     East Chelsea   
930  6939 Isabelle Station StApt. 204      3707  Andersonchester   

                  state    country  status  
602     South Australia  Australia  Active  
497  Northern Territory  Australia  Active  
341            Victoria  Australia  Active  
84   Northern Territory  Australia  Active  
930     New South Wales  Australia  Active

Kode di atas akan menghasilkan lima sample data, Bisa di ubah parameter 5 nya sesuai dengan keinginan untuk melihat data yg lebih banyak

Untuk memperoleh informasi terkait jumlah pelanggan yang berstatus “Active” dan “Non Active”, kita bisa menggunakan pivot table

In [ ]:
customers_df.groupby(by="status").customer_id.count()

status
Active        617
Non Active    384
Name: customer_id, dtype: int64

Seperti yang terlihat pada pivot table di atas, terdapat cukup banyak pelanggan yang belum pernah melakukan transaksi sebelumnya. Ini tentunya merupakan data yg buruk karena hampir 30% pelanggan belum pernah melakukan order sebelumnya.

Untuk memperoleh lebih banyak informasi terkait kedua data tersebut, kita perlu menggabungkan keduanya melalui proses join atau merge. Berikut merupakan contoh kode untuk melakukan merge terhadap data orders_df dan customers_df.

In [ ]:
orders_customers_df = pd.merge(
    left=orders_df,
    right=customers_df,
    how="left",
    left_on="customer_id",
    right_on="customer_id"
)
orders_customers_df.head()

order_id  customer_id  payment order_date delivery_date  delivery_time  \
0         1           64    30811 2021-08-30    2021-09-24           25.0   
1         2          473    50490 2021-02-03    2021-02-13           10.0   
2         3          774    46763 2021-10-08    2021-11-03           26.0   
3         4          433    39782 2021-05-06    2021-05-19           13.0   
4         5          441    14719 2021-03-23    2021-03-24            1.0   

  customer_name             gender  age                   home_address  \
0      fulan 64  Prefer not to say   75      4927 Alice MeadowApt. 960   
1     fulan 473               Male   61  531 Schmitt BoulevardApt. 010   
2     fulan 774  Prefer not to say   34       2096 Wilson MewsApt. 714   
3     fulan 433  Prefer not to say   46        5777 Mayer PassApt. 881   
4     fulan 441  Prefer not to say   53   33 Richards JunctionApt. 478   

   zip_code                city              state    country  status  
0      7787      Sanfordborough    South Australia  Australia  Active  
1      1744             Annaton    South Australia  Australia  Active  
2      8590  West Jonathanshire           Tasmania  Australia  Active  
3      9728    West Michaelport           Tasmania  Australia  Active  
4      7650     South Rileyview  Western Australia  Australia  Active

Pada kode di atas, kita melakukan proses merge dengan metode “left join”. metode ini memungkinkan kita untuk mengambil semua nilai dari tabel kiri serta nilai yang bersesuaian dengan tabel kanan.

Sehingga Terdapat banyak sekali informasi yang bisa kita eksplore dari data di atas. Berikut merupakan beberapa hal yang bisa dijelajahi.

Jumlah order berdasarkan kota Kita bisa membuat pivot table untuk memperoleh informasi terkait jumlah order berdasarkan kota dengan kode seperti berikut

In [ ]:
orders_customers_df.groupby(by="city").order_id.nunique().sort_values(ascending=False).reset_index().head(10)

city  order_id
0        Jordanside         6
1           New Ava         6
2  Port Hannahburgh         5
3         Lake Rose         5
4        O'keefeton         5
5          East Max         5
6          Rubyfort         5
7          West Kai         5
8         Sarahstad         4
9        Russelstad         4

berdasarkan hasil tersebut, diketahui bahwa kota Jordanside dan New Ava merupakan dua kota yang memiliki jumlah order terbanyak.

Jumlah order berdasarkan state Sekarang kita akan melihat jumlah order berdasarkan state. Untuk memperoleh informasi terkait hal ini, kita juga bisa menggunakan pivot table. Gunakan kode di bawah ini untuk menerapkannya.

In [ ]:
orders_customers_df.groupby(by="state").order_id.nunique().sort_values(ascending=False)

state
South Australia                 148
Queensland                      139
Western Australia               130
New South Wales                 129
Victoria                        118
Australian Capital Territory    118
Tasmania                        112
Northern Territory              106
Name: order_id, dtype: int64

Kode tersebut akan menghasilkan pivot table seperti di bawah ini. Berdasarkan pivot table tersebut, diketahui bahwa South Australia merupakan state yang paling banyak melakukan order.

Jumlah order berdasarkan gender Informasi selanjutnya yang dapat kita eksplorasi adalah jumlah order berdasarkan gender. Untuk melakukan hal ini, tentunya kita perlu membuat sebuah pivot table menggunakan kode berikut.

In [ ]:
orders_customers_df.groupby(by="gender").order_id.nunique().sort_values(ascending=False)

gender
Prefer not to say    716
Female               139
Male                 136
Name: order_id, dtype: int64

Berdasarkan hasil tersebut, dapat dilihat bahwa kebanyakan order dilakukan oleh pelanggan yang berjenis kelamin prefer not to say. Hal ini tentunya sejalan dengan jumlah pelanggan yang kebanyakan dari kelompok gender tersebut. Mengeksplorasi jumlah order berdasarkan kelompok usia. Untuk melakukan hal ini, kita perlu mendefinisikan sebuah kolom baru bernama “age_group”. Kolom ini akan membantu kita dalam mengelompokkan pelanggan ke dalam tiga kelompok, yaitu youth, adults, dan seniors. Selanjutnya, buatlah pivot table berdasarkan hal tersebut menggunakan contoh kode di bawah ini.

In [ ]:
orders_customers_df["age_group"] = orders_customers_df.age.apply(lambda x: "Youth" if x <= 24 else ("Seniors" if x > 64 else "Adults"))

In [ ]:
orders_customers_df.groupby(by="age_group").order_id.nunique().sort_values(ascending=False)

age_group
Adults     681
Seniors    226
Youth       93
Name: order_id, dtype: int64

Berdasarkan hasil tersebut diketahui bahwa pelanggan yang paling banyak melakukan order berasal dari kelompok usia Adults.Sejauh ini kita telah menemukan banyak sekali informasi yang tentunya menarik untuk ditelusuri lebih dalam. Namun, sebelum itu mari kita mengeksplorasi dua data yang lain (product_df dan sales_df) terlebih dahulu.Pada tahap ini, kita akan mengeksplorasi data product_df dan sales_df. Sebagai permulaan, kita dapat melihat rangkuman parameter statistik dari keduanya menggunakan method describe().

In [ ]:
products_df.describe(include="all")

product_id product_type product_name  size colour        price  \
count   1260.000000         1260         1260  1260   1260  1260.000000   
unique          NaN            3           35     8      7          NaN   
top             NaN        Shirt        Denim    XS    red          NaN   
freq            NaN          420           70   252    180          NaN   
mean     629.500000          NaN          NaN   NaN    NaN   105.805556   
std      363.874979          NaN          NaN   NaN    NaN     9.704423   
min        0.000000          NaN          NaN   NaN    NaN    90.000000   
25%      314.750000          NaN          NaN   NaN    NaN    95.750000   
50%      629.500000          NaN          NaN   NaN    NaN   108.500000   
75%      944.250000          NaN          NaN   NaN    NaN   114.000000   
max     1259.000000          NaN          NaN   NaN    NaN   119.000000   

           quantity                                        description  
count   1260.000000                                               1260  
unique          NaN                                               1260  
top             NaN  A violet coloured, XL sized, Tracksuit Bottoms...  
freq            NaN                                                  1  
mean      60.150000                                                NaN  
std       11.670573                                                NaN  
min       40.000000                                                NaN  
25%       50.000000                                                NaN  
50%       60.000000                                                NaN  
75%       70.000000                                                NaN  
max       80.000000                                                NaN

In [ ]:
sales_df.describe(include="all")

sales_id     order_id   product_id  price_per_unit    quantity  \
count  5000.000000  5000.000000  5000.000000     5000.000000  5000.00000   
mean   2499.500000   503.038200   634.053200      103.501600     1.99240   
std    1443.520003   285.964418   363.255794        9.195004     0.80751   
min       0.000000     1.000000     1.000000       90.000000     1.00000   
25%    1249.750000   258.000000   323.000000       95.000000     1.00000   
50%    2499.500000   504.500000   635.000000      102.000000     2.00000   
75%    3749.250000   749.000000   951.000000      112.000000     3.00000   
max    4999.000000   999.000000  1259.000000      119.000000     3.00000   

       total_price  
count  5000.000000  
mean    206.360000  
std      86.357457  
min      90.000000  
25%     112.000000  
50%     204.000000  
75%     285.000000  
max     357.000000

Berdasarkan hasil tersebut, dapat dilihat bahwa harga barang yang dijual berkisar antara 90 hingga 119 dolar. Selain itu, kita juga memperoleh informasi lain yang tidak kalah menarik yaitu pada setiap transaksi pelanggan paling banyak membeli tiga buah item dalam satu jenis produk dengan total harga sebesar 357 dolar. Apabila Anda penasaran dengan produk yang memiliki harga termahal dan terendah, silakan jalankan kode berikut untuk melihatnya.

In [ ]:
products_df.sort_values(by="price", ascending=False)

product_id product_type product_name size  colour  price  quantity  \
671          665       Jacket        Parka   XS     red    119        65   
672          666       Jacket        Parka    S     red    119        66   
673          667       Jacket        Parka    M     red    119        56   
674          668       Jacket        Parka    L     red    119        73   
675          669       Jacket        Parka   XL     red    119        64   
...          ...          ...          ...  ...     ...    ...       ...   
1222        1216     Trousers      Pleated    S  indigo     90        41   
1229        1223     Trousers      Pleated    L  violet     90        48   
1228        1222     Trousers      Pleated    M  violet     90        45   
1227        1221     Trousers      Pleated    S  violet     90        50   
1226        1220     Trousers      Pleated   XS  violet     90        58   

                                        description  
671          A red coloured, XS sized, Parka Jacket  
672           A red coloured, S sized, Parka Jacket  
673           A red coloured, M sized, Parka Jacket  
674           A red coloured, L sized, Parka Jacket  
675          A red coloured, XL sized, Parka Jacket  
...                                             ...  
1222   A indigo coloured, S sized, Pleated Trousers  
1229   A violet coloured, L sized, Pleated Trousers  
1228   A violet coloured, M sized, Pleated Trousers  
1227   A violet coloured, S sized, Pleated Trousers  
1226  A violet coloured, XS sized, Pleated Trousers  

[1260 rows x 8 columns]

Selanjutnya, kita bisa menggunakan pivot table untuk mencari informasi terkait produk berdasarkan tipe dan nama produknya. Berikut contoh kode yang bisa Anda gunakan.

In [ ]:
products_df.groupby(by="product_type").agg({
    "product_id": "nunique",
    "quantity": "sum",
    "price":  ["min", "max"]
})

product_id quantity price     
                nunique      sum   min  max
product_type                               
Jacket              420    25387    90  119
Shirt               420    25263    92  119
Trousers            420    25139    90  119

In [ ]:
products_df.groupby(by="product_name").agg({
    "product_id": "nunique",
    "quantity": "sum",
    "price": ["min", "max"]
})

product_id quantity price     
                     nunique      sum   min  max
product_name                                    
Bomber                    35     2083    90   90
Camp Collared             35     2071   112  112
Cardigan                  35     2032   118  118
Cargo Pants               35     2146   106  106
Casual Slim Fit           35     2086   113  113
Chambray                  35     2020   105  105
Chinos                    35     2101   100  100
Coach                     35     2158   115  115
Cords                     35     2260   113  113
Cropped                   35     2085    99   99
Cuban Collar              35     2017    93   93
Denim                     70     4204    92  115
Drawstring                35     2042   104  104
Dress                     35     2125   119  119
Flannel                   35     2135    96   96
Henley                    35     2051    92   92
High-Waisted              35     2198    98   98
Joggers                   35     2107    94   94
Leather                   35     2276   113  113
Linen                     35     2138   116  116
Mandarin Collar           35     2203   108  108
Oxford Cloth              35     2071   114  114
Parka                     35     2201   119  119
Peacoat                   35     2067   102  102
Pleated                   35     2030    90   90
Polo                      35     2196   117  117
Puffer                    35     1946   110  110
Pullover                  35     2184   114  114
Relaxed Leg               35     2002    95   95
Shearling                 35     2169    95   95
Slim-Fit                  35     2155   119  119
Tracksuit Bottoms         35     2038    91   91
Trench Coat               35     2132   112  112
Windbreaker               35     2085   109  109
Wool                      35     1975   111  111

Pivot table di atas dapat memberikan kita gambaran terkait produk yang dijual oleh Fashion Collection.Sebagai calon praktisi data yang Andal, tentunya Anda penasaran dengan produk yang paling laris. Nah, untuk menjawab pertanyaan ini, kita perlu menyatukan (merge) tabel product_df dan sales_df dengan kode berikut.

In [ ]:
sales_products_df = pd.merge(
    left=sales_df,
    right=products_df,
    how="left",
    left_on="product_id",
    right_on="product_id"
)
sales_products_df.head()

sales_id  order_id  product_id  price_per_unit  quantity_x  total_price  \
0         0         1         218             106           2          212   
1         1         1         481             118           1          118   
2         2         1           2              96           3          288   
3         3         1        1002             106           2          212   
4         4         1         691             113           3          339   

  product_type  product_name size  colour  price  quantity_y  \
0        Shirt      Chambray    L  orange    105          44   
1       Jacket        Puffer    S  indigo    110          62   
2        Shirt  Oxford Cloth    M     red    114          54   
3     Trousers          Wool    M    blue    111          52   
4       Jacket         Parka    S  indigo    119          53   

                                   description  
0   A orange coloured, L sized, Chambray Shirt  
1    A indigo coloured, S sized, Puffer Jacket  
2  A red coloured, M sized, Oxford Cloth Shirt  
3      A blue coloured, M sized, Wool Trousers  
4     A indigo coloured, S sized, Parka Jacket

Pada dasarnya kode di atas akan melakukan proses merge terhadap data product_df dan sales_df. Proses merge tersebut dilakukan menggunakan metode “left”. Berikut hasil dari proses merge tersebut.

Jika diperhatikan kembali, hasil dari proses merge di atas memiliki perbedaan antara nilai price_per_unit dan price. Hal ini bisa terjadi karena adanya potongan harga, biaya operasional, dan biaya lainnya.

sekarang kita coba melihat informasi penjualan produk berdasarkan tipenya. dengan membuat pivot table berdasarkan produk type dengan kode berikut

In [ ]:
sales_products_df.groupby(by="product_type").agg({
    "sales_id": "nunique",
    "quantity_x": "sum",
    "total_price": "sum"
})

sales_id  quantity_x  total_price
product_type                                   
Jacket            1676        3343       357026
Shirt             1641        3259       333600
Trousers          1683        3360       341174


Jika Anda perhatikan pivot table di atas, Trousers merupakan tipe produk yang paling laris. Akan tetapi, jika dilihat berdasarkan revenue yang diterima, Jacket merupakan tipe produk yang paling banyak menyumbang revenue perusahaan.

membuat pivot table yang sama untuk melihat informasi penjualan berdasarkan nama produk menggunakan kode di bawah ini


     

In [ ]:
sales_products_df.groupby(by="product_name").agg({
    "sales_id": "nunique",
    "quantity_x": "sum",
    "total_price": "sum"
}).sort_values(by="total_price", ascending=False)

sales_id  quantity_x  total_price
product_name                                        
Denim                   273         527        52399
Casual Slim Fit         154         306        36414
Trench Coat             146         299        35581
Shearling               150         302        35334
Puffer                  140         298        35164
Flannel                 141         281        33158
Cropped                 135         284        32660
Pleated                 147         308        32340
Joggers                 164         334        31062
Chambray                141         290        30740
Parka                   134         269        30397
Bomber                  150         297        29700
Chinos                  135         260        29380
Windbreaker             143         287        29274
Cardigan                134         260        29120
Peacoat                 132         266        28994
Tracksuit Bottoms       130         257        28784
Slim-Fit                154         288        27936
Oxford Cloth            146         289        27744
Drawstring              147         288        27648
High-Waisted            140         278        27522
Camp Collared           139         279        27342
Cargo Pants             134         267        27234
Relaxed Leg             146         301        27090
Cuban Collar            122         254        26670
Henley                  146         281        26414
Wool                    125         249        26394
Leather                 136         263        25774
Mandarin Collar         124         236        25724
Dress                   127         243        25272
Pullover                143         276        25116
Polo                    131         264        25080
Coach                   134         265        24645
Linen                   131         270        24570
Cords                   126         246        23124

Berdasarkan pivot table tersebut, diketahui bahwa produk Denim merupakan produk yang paling laris dan juga paling banyak menyumbang revenue terhadap perusahaan.

kita perlu membuat sebuah DataFrame baru bernama all_df untuk menampung semua informasi dari keempat tabel yang kita miliki. Berikut contoh kode yang dapat kita gunakan untuk menjalankan proses penggabungan ini.

In [ ]:
import pandas as pd
all_df = pd.merge(
    left=sales_products_df,
    right=orders_customers_df,
    how="left",
    left_on="order_id",
    right_on="order_id"
)
all_df.head()

sales_id  order_id  product_id  price_per_unit  quantity_x  total_price  \
0         0         1         218             106           2          212   
1         1         1         481             118           1          118   
2         2         1           2              96           3          288   
3         3         1        1002             106           2          212   
4         4         1         691             113           3          339   

  product_type  product_name size  colour  ...  customer_name  \
0        Shirt      Chambray    L  orange  ...       fulan 64   
1       Jacket        Puffer    S  indigo  ...       fulan 64   
2        Shirt  Oxford Cloth    M     red  ...       fulan 64   
3     Trousers          Wool    M    blue  ...       fulan 64   
4       Jacket         Parka    S  indigo  ...       fulan 64   

              gender age               home_address  zip_code            city  \
0  Prefer not to say  75  4927 Alice MeadowApt. 960      7787  Sanfordborough   
1  Prefer not to say  75  4927 Alice MeadowApt. 960      7787  Sanfordborough   
2  Prefer not to say  75  4927 Alice MeadowApt. 960      7787  Sanfordborough   
3  Prefer not to say  75  4927 Alice MeadowApt. 960      7787  Sanfordborough   
4  Prefer not to say  75  4927 Alice MeadowApt. 960      7787  Sanfordborough   

             state    country  status age_group  
0  South Australia  Australia  Active   Seniors  
1  South Australia  Australia  Active   Seniors  
2  South Australia  Australia  Active   Seniors  
3  South Australia  Australia  Active   Seniors  
4  South Australia  Australia  Active   Seniors  

[5 rows x 28 columns]

Proses merge di atas akan menghasilkan DataFrame seperti hasil di atas

Selanjutnya, mari kita coba lihat preferensi pembelian berdasarkan state pelanggan dan tipe produk menggunakan kode di bawah ini.

In [ ]:
all_df.groupby(by=["state", "product_type"]).agg({
    "quantity_x": "sum",
    "total_price": "sum"
})

quantity_x  total_price
state                        product_type                         
Australian Capital Territory Jacket               406        43204
                             Shirt                396        40448
                             Trousers             454        46790
New South Wales              Jacket               451        47998
                             Shirt                431        43980
                             Trousers             392        39766
Northern Territory           Jacket               365        38991
                             Shirt                336        33865
                             Trousers             384        38998
Queensland                   Jacket               499        53511
                             Shirt                417        42506
                             Trousers             453        46045
South Australia              Jacket               461        49090
                             Shirt                509        52685
                             Trousers             455        46041
Tasmania                     Jacket               412        44370
                             Shirt                387        39668
                             Trousers             379        38057
Victoria                     Jacket               359        38203
                             Shirt                390        40206
                             Trousers             414        41948
Western Australia            Jacket               390        41659
                             Shirt                393        40242
                             Trousers             429        43529

Pivot table di atas memberikan kita gambaran terkait tipe produk yang disukai pengguna berdasarkan lokasi negara bagiannya. Jika diperhatikan, tipe produk Jacket (garis merah) paling banyak terjual pada negara bagian Queensland, South Australia, dan New South Wales. Untuk tipe produk lain, Anda bisa melihatnya pada gambar pivot table di atas.

Sekarang Anda bisa melakukan hal yang sama untuk mengetahui selera tipe produk pelanggan berdasarkan gender dan kelompok usia. Berikut contoh kode yang bisa Anda gunakan.

In [ ]:
all_df.groupby(by=["gender", "product_type"]).agg({
    "quantity_x": "sum",
    "total_price": "sum"
})

quantity_x  total_price
gender            product_type                         
Female            Jacket               481        50963
                  Shirt                445        45146
                  Trousers             481        49295
Male              Jacket               480        50978
                  Shirt                410        41939
                  Trousers             453        46071
Prefer not to say Jacket              2349       251486
                  Shirt               2386       244765
                  Trousers            2400       243092

In [ ]:
all_df.groupby(by=["age_group", "product_type"]).agg({
    "quantity_x": "sum",
    "total_price": "sum"
})

quantity_x  total_price
age_group product_type                         
Adults    Jacket              2292       245055
          Shirt               2225       227781
          Trousers            2272       231271
Seniors   Jacket               777        82959
          Shirt                733        74977
          Trousers             766        77471
Youth     Jacket               274        29012
          Shirt                301        30842
          Trousers             322        32432

Pivot table di atas dapat memberikan kita gambaran terkait selera tipe produk pelanggan berdasarkan gender dan kelompok usia. Semua informasi ini akan sangat membantu Anda dalam menjawab pertanyaan terkait preferensi pengguna berdasarkan demografinya.



In [3]:
import pandas as pd
customers =pd.read_csv('/content/drive/MyDrive/Pemrograman Dasar/Pertemuan 10/customers.csv')
orders =pd.read_csv('/content/drive/MyDrive/Pemrograman Dasar/Pertemuan 10/orders.csv')
products =pd.read_csv('/content/drive/MyDrive/Pemrograman Dasar/Pertemuan 10/products.csv')
sales =pd.read_csv('/content/drive/MyDrive/Pemrograman Dasar/Pertemuan 10/sales.csv')

In [4]:
orders['order_date'] = pd.to_datetime(orders['order_date'])

In [9]:
sales_orders_df = sales_products_df.merge(orders[['order_id', 'order_date']], on= 'order_id', how='left')

In [10]:
sales_orders_df['renevue']=(sales_orders_df['price_per_unit'] * sales_orders_df['quantity_x'])

In [11]:
sales_orders_df['month'] = sales_orders_df['order_date'].dt.to_period('M')

In [12]:
monthly_performance = sales_orders_df.groupby('month').agg(total_quantity=('quantity_x', 'sum'),total_renevue= ('renevue','sum')).reset_index()

Untuk menghitung performa penjualan dan renevue perbulan

In [13]:
monthly_performance

month  total_quantity  total_renevue
0  2021-01            1148         119851
1  2021-02             924          95480
2  2021-03            1266         131600
3  2021-04             951          98230
4  2021-05             840          86631
5  2021-06             985         101762
6  2021-07            1122         116081
7  2021-08            1012         105019
8  2021-09             947          97608
9  2021-10             819          84734

Output yang ditampilkan setelah dijalankan.

In [14]:
sales_products_df = sales.merge(products, on='product_id', how='left')

Untuk menggabungkan data sales dan products

In [15]:
sales_products_df.columns


Index(['sales_id', 'order_id', 'product_id', 'price_per_unit', 'quantity_x',
       'total_price', 'product_type', 'product_name', 'size', 'colour',
       'price', 'quantity_y', 'description'],
      dtype='object')

Untuk menghitung total penjualan perproduk

In [16]:
product_sale = sales_products_df.groupby('product_name')['quantity_x'].sum().reset_index()
product_sale

product_name  quantity_x
0              Bomber         321
1       Camp Collared         279
2            Cardigan         260
3         Cargo Pants         267
4     Casual Slim Fit         306
5            Chambray         290
6              Chinos         260
7               Coach         265
8               Cords         246
9             Cropped         284
10       Cuban Collar         254
11              Denim         527
12         Drawstring         288
13              Dress         245
14            Flannel         281
15             Henley         283
16       High-Waisted         278
17            Joggers         334
18            Leather         263
19              Linen         286
20    Mandarin Collar         236
21       Oxford Cloth         289
22              Parka         269
23            Peacoat         266
24            Pleated         308
25               Polo         264
26             Puffer         306
27           Pullover         276
28        Relaxed Leg         301
29          Shearling         302
30           Slim-Fit         288
31  Tracksuit Bottoms         257
32        Trench Coat         299
33        Windbreaker         287
34               Wool         249

Untuk menghitung total penjualan Setiap produk

In [17]:
produk_terlaris = product_sale.sort_values(by='quantity_x', ascending=False)
produk_terlaris

product_name  quantity_x
11              Denim         527
17            Joggers         334
0              Bomber         321
24            Pleated         308
4     Casual Slim Fit         306
26             Puffer         306
29          Shearling         302
28        Relaxed Leg         301
32        Trench Coat         299
5            Chambray         290
21       Oxford Cloth         289
12         Drawstring         288
30           Slim-Fit         288
33        Windbreaker         287
19              Linen         286
9             Cropped         284
15             Henley         283
14            Flannel         281
1       Camp Collared         279
16       High-Waisted         278
27           Pullover         276
22              Parka         269
3         Cargo Pants         267
23            Peacoat         266
7               Coach         265
25               Polo         264
18            Leather         263
2            Cardigan         260
6              Chinos         260
31  Tracksuit Bottoms         257
10       Cuban Collar         254
34               Wool         249
8               Cords         246
13              Dress         245
20    Mandarin Collar         236

Untuk menampilkan produk yang paling banyak terjual

In [18]:
produk_paling_sedikit = product_sale.sort_values(by='quantity_x', ascending=True)
produk_paling_sedikit

product_name  quantity_x
20    Mandarin Collar         236
13              Dress         245
8               Cords         246
34               Wool         249
10       Cuban Collar         254
31  Tracksuit Bottoms         257
6              Chinos         260
2            Cardigan         260
18            Leather         263
25               Polo         264
7               Coach         265
23            Peacoat         266
3         Cargo Pants         267
22              Parka         269
27           Pullover         276
16       High-Waisted         278
1       Camp Collared         279
14            Flannel         281
15             Henley         283
9             Cropped         284
19              Linen         286
33        Windbreaker         287
12         Drawstring         288
30           Slim-Fit         288
21       Oxford Cloth         289
5            Chambray         290
32        Trench Coat         299
28        Relaxed Leg         301
29          Shearling         302
4     Casual Slim Fit         306
26             Puffer         306
24            Pleated         308
0              Bomber         321
17            Joggers         334
11              Denim         527

Untuk menampilkan paling sedikit terjual

In [20]:
print("Produk paling banyak terjual:")
print(produk_terlaris.iloc[0])

print("\nProduk paling sedikit terjual:")
print(produk_paling_sedikit.iloc[0])

Produk paling banyak terjual:
product_name    Denim
quantity_x        527
Name: 11, dtype: object

Produk paling sedikit terjual:
product_name    Mandarin Collar
quantity_x                  236
Name: 20, dtype: object


untuk menampilkan 1 produk yang terbanyak terjual dan yang paling sedikit terjual

In [21]:
customers.columns

Index(['customer_id', 'customer_name', 'gender', 'age', 'home_address',
       'zip_code', 'city', 'state', 'country'],
      dtype='object')

Untuk mengecek kolom nya

In [22]:
demografi_gender = customers.groupby('gender').size().reset_index(name='total_pelanggan')
demografi_gender

gender  total_pelanggan
0             Female              115
1               Male              143
2  Prefer not to say              731

untuk mengetahui jumlah gender yang ada pada data

In [28]:
customers.columns

Index(['customer_id', 'customer_name', 'gender', 'age', 'home_address',
       'zip_code', 'city', 'state', 'country'],
      dtype='object')

untuk mengecek kolom dalam data

In [29]:
customers['age_group'] = pd.cut(
    customers['age'],
    bins=[0, 24, 64, 100],
    labels=['Youth', 'Adults', 'Seniors'],
    right=True
)

In [52]:
demografi_usia = customers.groupby('age_group').size().reset_index(name='total_pelanggan')
demografi_usia

/tmp/ipython-input-3553882017.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  demografi_usia = customers.groupby('age_group').size().reset_index(name='total_pelanggan')


age_group  total_pelanggan
0     Youth               84
1    Adults              670
2   Seniors              251

Untuk membuat kelompok usia

In [30]:
customers[['age','age_group']].head()

age age_group
0   30    Adults
1   69   Seniors
2   59    Adults
3   67   Seniors
4   30    Adults

Untuk mengecek hasilnya

In [31]:
demografi_usia = customers.groupby('age_group').size().reset_index(name='total_pelanggan')
demografi_usia

/tmp/ipython-input-3553882017.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  demografi_usia = customers.groupby('age_group').size().reset_index(name='total_pelanggan')


age_group  total_pelanggan
0     Youth               84
1    Adults              670
2   Seniors              251

demografi pelanggan berdasarkan usia

In [32]:
demografi_gender = customers.groupby('gender').size().reset_index(name='total_pelanggan')
demografi_gender

gender  total_pelanggan
0             Female              115
1               Male              143
2  Prefer not to say              731

demografi berdasarkan gender

In [33]:
demografi_usia_gender = pd.pivot_table(
    customers,
    index='age_group',
    columns='gender',
    aggfunc='size',
    fill_value=0
)

demografi_usia_gender


/tmp/ipython-input-3371669982.py:1: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  demografi_usia_gender = pd.pivot_table(


gender     Female  Male  Prefer not to say
age_group                                 
Youth          11    14                 58
Adults         83    87                486
Seniors        21    42                185

Untuk mendapatkan data yang lebih lengkap

In [34]:
orders['order_date'] = pd.to_datetime(orders['order_date'])

Untuk mendapatkan data order date

In [35]:
sales_orders_df = sales.merge(
    orders[['order_id', 'customer_id', 'order_date']],
    on='order_id',
    how='left'
)

setelah itu kita menggabungkan data transaksi dan daat order

In [53]:
last_transaction = sales_orders_df.groupby('customer_id')['order_date'].max().reset_index()
last_transaction

customer_id order_date
0              1 2021-04-04
1              7 2021-05-21
2             10 2021-03-09
3             11 2021-05-28
4             12 2021-06-19
..           ...        ...
611          994 2021-01-12
612          995 2021-08-02
613          996 2021-09-03
614          998 2021-07-19
615         1000 2021-10-11

[616 rows x 2 columns]

Melihat transaksi terakhir tiap pelanggan

In [37]:
last_transaction_detail = last_transaction.merge(
    customers[['customer_id', 'customer_name']],
    on='customer_id',
    how='left'
)

last_transaction_detail


customer_id order_date customer_name
0              1 2021-04-04       fulan 1
1              7 2021-05-21       fulan 7
2             10 2021-03-09      fulan 10
3             11 2021-05-28      fulan 11
4             12 2021-06-19      fulan 12
..           ...        ...           ...
616          994 2021-01-12     fulan 994
617          995 2021-08-02     fulan 995
618          996 2021-09-03     fulan 996
619          998 2021-07-19     fulan 998
620         1000 2021-10-11    fulan 1000

[621 rows x 3 columns]

Setelah itu kita menggabungkan nama pelanggan

In [38]:
last_transaction_detail.sort_values(by='order_date')

customer_id order_date customer_name
164          276 2021-01-01     fulan 276
125          212 2021-01-02     fulan 212
589          946 2021-01-02     fulan 946
417          676 2021-01-02     fulan 676
492          790 2021-01-03     fulan 790
..           ...        ...           ...
250          415 2021-10-23     fulan 415
98           163 2021-10-23     fulan 163
427          687 2021-10-24     fulan 687
428          690 2021-10-24     fulan 690
241          401 2021-10-24     fulan 401

[621 rows x 3 columns]

ini adalah yang paling lama tidak transaksi

In [40]:
orders['order_date'] = pd.to_datetime(orders['order_date'])

masukan data order date pada google colab

In [41]:
purchase_frequency = orders.groupby('customer_id')['order_id'].nunique().reset_index()
purchase_frequency.rename(columns={'order_id': 'total_pembelian'}, inplace=True)

purchase_frequency


customer_id  total_pembelian
0              1                3
1              7                1
2             10                1
3             11                1
4             12                1
..           ...              ...
612          994                1
613          995                1
614          996                2
615          998                4
616         1000                1

[617 rows x 2 columns]

Hitung frekuensi pembelian perpelanggan

In [42]:
purchase_frequency_detail = purchase_frequency.merge(
    customers[['customer_id', 'customer_name']],
    on='customer_id',
    how='left'
)

purchase_frequency_detail


customer_id  total_pembelian customer_name
0              1                3       fulan 1
1              7                1       fulan 7
2             10                1      fulan 10
3             11                1      fulan 11
4             12                1      fulan 12
..           ...              ...           ...
617          994                1     fulan 994
618          995                1     fulan 995
619          996                2     fulan 996
620          998                4     fulan 998
621         1000                1    fulan 1000

[622 rows x 3 columns]

dan gabungkan nama pelanggan

In [43]:
purchase_frequency_detail.sort_values(by='total_pembelian', ascending=False)

customer_id  total_pembelian customer_name
352          571                6     fulan 571
348          566                5     fulan 566
212          348                5     fulan 348
166          282                5     fulan 282
412          664                5     fulan 664
..           ...              ...           ...
601          966                1     fulan 966
602          969                1     fulan 969
603          970                1     fulan 970
605          973                1     fulan 973
12            22                1      fulan 22

[622 rows x 3 columns]

Untuk mencari nama pelanggan yang rajin membeli


In [44]:
purchase_frequency_detail['kategori_frekuensi'] = pd.cut(
    purchase_frequency_detail['total_pembelian'],
    bins=[0, 1, 3, 100],
    labels=['Rendah', 'Sedang', 'Tinggi']
)

purchase_frequency_detail


customer_id  total_pembelian customer_name kategori_frekuensi
0              1                3       fulan 1             Sedang
1              7                1       fulan 7             Rendah
2             10                1      fulan 10             Rendah
3             11                1      fulan 11             Rendah
4             12                1      fulan 12             Rendah
..           ...              ...           ...                ...
617          994                1     fulan 994             Rendah
618          995                1     fulan 995             Rendah
619          996                2     fulan 996             Sedang
620          998                4     fulan 998             Tinggi
621         1000                1    fulan 1000             Rendah

[622 rows x 4 columns]

Untuk mengelompokan frekuensi yang paling rendah sampai tertinggi

In [45]:
sales_orders_df = sales_products_df.merge(
    orders[['order_id', 'customer_id', 'order_date']],
    on='order_id',
    how='left'
)

untuk mendapatkan data order date dan bisa muncul pada coolab ketika di panggil


In [46]:
sales_orders_df['revenue'] = (
    sales_orders_df['price_per_unit'] * sales_orders_df['quantity_x']
)

untuk menghitung data renevue per transaksi

In [47]:
customer_spending = sales_orders_df.groupby('customer_id')['revenue'].sum().reset_index()
customer_spending.rename(columns={'revenue': 'total_uang_dihabiskan'}, inplace=True)

customer_spending


customer_id  total_uang_dihabiskan
0              1                   1641
1              7                   1017
2             10                    270
3             11                    382
4             12                   1551
..           ...                    ...
611          994                   1716
612          995                   2140
613          996                   3045
614          998                   4415
615         1000                   1483

[616 rows x 2 columns]

Untuk mendapatkan total uang yang dihabisin tiap pelanggan

In [49]:
customer_spending_detail = customer_spending.merge(
    customers[['customer_id', 'customer_name']],
    on='customer_id',
    how='left'
)

customer_spending_detail

customer_id  total_uang_dihabiskan customer_name
0              1                   1641       fulan 1
1              7                   1017       fulan 7
2             10                    270      fulan 10
3             11                    382      fulan 11
4             12                   1551      fulan 12
..           ...                    ...           ...
616          994                   1716     fulan 994
617          995                   2140     fulan 995
618          996                   3045     fulan 996
619          998                   4415     fulan 998
620         1000                   1483    fulan 1000

[621 rows x 3 columns]

setelah itu kita gabungkan tiap tiap nama orang

In [50]:
customer_spending_detail.sort_values(by='total_uang_dihabiskan', ascending=False)

customer_id  total_uang_dihabiskan customer_name
166          282                   7632     fulan 282
351          571                   6640     fulan 571
411          664                   6341     fulan 664
120          206                   5881     fulan 206
347          566                   5671     fulan 566
..           ...                    ...           ...
547          874                    200     fulan 874
601          969                    186     fulan 969
26            47                    183      fulan 47
514          818                    112     fulan 818
502          802                    100     fulan 802

[621 rows x 3 columns]

Untuk mendapatkan pelanggan yang paling boros :)

KODE PROGRAM TUGAS 10
DONE:)